src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java

```diff
@@ -0,0 +1,313 @@
+		this.client.setBlocking(!stopOnEOF);
+		this.client.registerEventListener(binlogEventListener);
+		this.client.setServerId(replicaServerID.intValue());
+		this.client.setBinlogFilename(start.getFile());
+		this.client.setBinlogPosition(start.getOffset());
+
+		/*
+		this.shouldHeartbeat = shouldHeartbeat;
+		if ( shouldHeartbeat )
+			this.replicator.setHeartbeatPeriod(0.5f);
+			*/
+
+		this.stopOnEOF = stopOnEOF;
+
+		this.maxwellSchemaDatabaseName = maxwellSchemaDatabaseName;
+		this.setBinlogPosition(start);
+	}
+
+	public BinlogConnectorReplicator(SchemaStore schemaStore, AbstractProducer producer, AbstractBootstrapper bootstrapper, MaxwellContext ctx, BinlogPosition start) throws SQLException {
+		this(
+			schemaStore,
+			producer,
+			bootstrapper,
+			ctx.getConfig().replicationMysql,
+			ctx.getConfig().replicaServerID,
+			ctx.shouldHeartbeat(),
+			ctx.getPositionStoreThread(),
+			ctx.getConfig().databaseName,
+			start,
+			false,
+			ctx.getConfig().clientID
+		);
+	}
+
+	public void setBinlogPosition(BinlogPosition p) {
+		this.client.setBinlogFilename(p.getFile());
+		this.client.setBinlogPosition(p.getOffset());
+	}
+
+	private void ensureReplicatorThread() throws Exception {
+		if ( !client.isConnected() && !stopOnEOF ) {
+			LOGGER.warn("replicator stopped at position " + client.getBinlogFilename() + ":" + client.getBinlogPosition() + " -- restarting");
+			client.connect(5000);
+		}
+	}
+
+	public void startReplicator() throws Exception {
+		this.client.connect(5000);
+	}
+
+	@Override
+	protected void beforeStart() throws Exception {
+		startReplicator();
+	}
+
+	@Override
+	protected void beforeStop() throws Exception {
+		this.binlogEventListener.stop();
+		this.client.disconnect();
+	}
+
+	private static Pattern createTablePattern =
+			Pattern.compile("^CREATE\\s+TABLE", Pattern.CASE_INSENSITIVE);
+
+	/**
+	 * Get a batch of rows for the current transaction.
+	 *
+	 * We assume the replicator has just processed a "BEGIN" event, and now
+	 * we're inside a transaction.  We'll process all rows inside that transaction
+	 * and turn them into RowMap objects.  We do this because mysql attaches the
+	 * transaction-id (xid) to the COMMIT event (at the end of the transaction),
+	 * so we process the entire transaction in order to assign each row the same xid.
+
+	 * @return A RowMapBuffer of rows; either in-memory or on disk.
+	 */
+
+	private RowMapBuffer getTransactionRows() throws Exception {
+		BinlogConnectorEvent event;
+		RowMapBuffer buffer = new RowMapBuffer(MAX_TX_ELEMENTS);
+
+		while ( true ) {
+			event = pollEvent();
+
+			if (event == null) {
+				ensureReplicatorThread();
+				continue;
+			}
+
+			switch(event.getEvent().getHeader().getEventType()) {
+				case WRITE_ROWS:
+				case UPDATE_ROWS:
+				case DELETE_ROWS:
+				case EXT_WRITE_ROWS:
+				case EXT_UPDATE_ROWS:
+				case EXT_DELETE_ROWS:
+					Table table = tableCache.getTable(event.getTableID());
+					event.setLastHeartbeat(lastHeartbeatRead);
```

Comment:
I think I get this now, the old replicator would instantiate subclasses of `AbstractRowsEvent` with the `lastHeartbeatRead`, so when building RowMap's it was already present; but now the equivalent `BinlogConnectorEvent` deals with types (e.g. update) itself, so it needs lastHeartbeatRead set, so it can instantiate RowMap's.